### ISSUE TITLE
Support for writable uniform image buffers in compute shaders

### ISSUE THREAD

**Issue by portaloffreedom** [EmbarkStudios/rust-gpu#886]
Hi! 

I'm new to the gpu low-level space, so I hope I don't say something extremely stupid.
I'm trying to follow the vulkano.rs tutorial but replacing the glsl code with rust. At the beginning it starts with compute shaders before doing vertex or fragment shaders, which I understand are the higher first priority for this project and compute is a low priority.

The [tutorial](https://vulkano.rs/guide/mandelbrot) is trying to make a compute shader that writes a mandlebrot fractal into an image.
The image is written to as a  writeonly 2d image buffer.
```glsl
layout(set = 0, binding = 0, rgba8) uniform writeonly image2D img;
```

The corresponding code, as far as I can tell, would be 
```rust
type Image2d = Image!(2D, type=f32, sampled=false, depth=false);

#[spirv(compute(threads(8,8)))]
pub fn fractal(
    #[spirv(global_invocation_id)] id: UVec3,
    #[spirv(uniform, descriptor_set = 0, binding = 0)] image: &mut Image2d,
) { ... }
```
but when compiling it complains about the inferred type being different from the attribute specified type
```
     Compiling fractal v0.1.0 (/home/matteo/projects/rust/rustgpu-test/shaders/fractal)
  error: storage class mismatch
    --> shaders/fractal/src/lib.rs:22:5
     |
  22 |     #[spirv(uniform, descriptor_set = 0, binding = 0)] image: &mut Image2d,
     |     ^^^^^^^^-------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^------------
     |             |                                                 |
     |             Uniform specified in attribute                    UniformConstant inferred from type
     |
  help: remove storage class attribute to use UniformConstant as storage class
    --> shaders/fractal/src/lib.rs:22:13
     |
  22 |     #[spirv(uniform, descriptor_set = 0, binding = 0)] image: &mut Image2d,
     |             ^^^^^^^

  error: could not compile `fractal` due to previous error
```

I think I'm not allowed at all to have writable images, based on the code here:
https://github.com/EmbarkStudios/rust-gpu/blob/132b960e8320d7e607fd912142cff7da05d49d92/crates/rustc_codegen_spirv/src/codegen_cx/entry.rs#L190-L209

Since glsl can support this, what is stopping rust-gpu to support it as well? Would it be possible to support writable images on rust-gpu?

**Comment by msiglreith** [comment]
The `uniform` in glsl here doesn't has the same semantic as the specified `uniform` storage class in the `spirv(..)` attribute. In GLSL it refers to the variable being uniform across invocations while the storage class refers is an detail of SPIRV/Vulkan. Image has `image` storage class which is automatically inferred by the compiler. The following snippet should work. `Image` type also don't need to be specified as `&mut`.
```rust
#[spirv(descriptor_set = 0, binding = 0)] image: &Image2d
```

**Comment by portaloffreedom** [comment]
You are right! It does work (it requires me to set the format of the image in the shader, but the compiler error was clear enough for that). It is a bit unintuitive that you can write on a non mutable UniformConstant parameter, food for thought if you want to improve the interface ;)
